Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp

```diff
@@ -83,20 +85,42 @@ class RemoteImageBufferSetProxyFlushFence : public ThreadSafeRefCounted<RemoteIm
 
 namespace {
 
-class RemoteImageBufferSetProxyFlusher final : public ThreadSafeImageBufferFlusher {
+class RemoteImageBufferSetProxyFlusher final : public ThreadSafeImageBufferSetFlusher {
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    RemoteImageBufferSetProxyFlusher(Ref<RemoteImageBufferSetProxyFlushFence> flushState)
-        : m_flushState(WTFMove(flushState))
+    RemoteImageBufferSetProxyFlusher(RemoteImageBufferSetIdentifier identifier, Ref<RemoteImageBufferSetProxyFlushFence> flushState, Ref<RemoteRenderingBackendProxyBackground> background, unsigned generation)
```

Comment:
This flush fence does now two things:
- wait until flush has completed
- wait until the handles have arrived.

The currently matching implementation for RemoteImageBufferSetProxy::didPrepareForDisplay would be:
```
class RemoteImageBufferSetProxyFlushFence ... { 
    bool waitFor(Seconds timeout)
    {
        Locker locker { m_lock };
        if (!m_signaled)
            m_signaled = m_event.waitFor(timeout);
        if (m_signaled)
            tracePoint(FlushRemoteImageBufferEnd, reinterpret_cast<uintptr_t>(this), 0u);
        if (m_signaled &&!m_backendHandles)
           m_condition.waitUntil(...);
        return m_signaled && m_backendHandles;
    }
    void setBackendHandles(..) {
        Locker locker { m_lock };
        m_backendHandles = ...;
        m_condition.signalAll(); 
      )
};

void RemoteImageBufferSetProxy::didPrepareForDisplay(outputdata) (
    if (!m_pendingFlush)
       return;
    m_pendingFlush->setBackendHandles(outputData);
}
```


E.g. store the backend handles that need to be waited in the wait object shared state.

however: the code doesn't handle the case where wait for the handles or flush times out, the code progresses and flushes again, and then two flushes complete: the old timed out one and the new correct one..

Reply:
That last bit could indeed be a problem. I might need to add an identifier to the prepareForDisplay call, returning into the didPrepareForDisplay response, to identify if it's the one we're currently waiting for.